Content.Server/Revenant/EntitySystems/RevenantSystem.cs

```diff
@@ -0,0 +1,198 @@
+{
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly ActionsSystem _action = default!;
+    [Dependency] private readonly AlertsSystem _alerts = default!;
+    [Dependency] private readonly DamageableSystem _damage = default!;
+    [Dependency] private readonly DoAfterSystem _doAfter = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly PolymorphableSystem _polymorphable = default!;
+    [Dependency] private readonly PhysicsSystem _physics = default!;
+    [Dependency] private readonly StatusEffectsSystem _statusEffects = default!;
+    [Dependency] private readonly SharedInteractionSystem _interact = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedStunSystem _stun = default!;
+    [Dependency] private readonly TagSystem _tag = default!;
+    [Dependency] private readonly MovementSpeedModifierSystem _movement = default!;
+    [Dependency] private readonly VisibilitySystem _visibility = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<RevenantComponent, ComponentStartup>(OnInit);
+
+        SubscribeLocalEvent<RevenantComponent, DamageChangedEvent>(OnDamage);
+        SubscribeLocalEvent<RevenantComponent, ExaminedEvent>(OnExamine);
+        SubscribeLocalEvent<RevenantComponent, StatusEffectAddedEvent>(OnStatusAdded);
+        SubscribeLocalEvent<RevenantComponent, StatusEffectEndedEvent>(OnStatusEnded);
+
+        InitializeAbilities();
+        InitializeShop();
+    }
+
+    private void OnInit(EntityUid uid, RevenantComponent component, ComponentStartup args)
+    {
+        //update the icon
+        ChangeEssenceAmount(uid, 0, component);
+
+        //default the visuals
+        if (TryComp<AppearanceComponent>(uid, out var app))
+        {
+            app.SetData(RevenantVisuals.Corporeal, false);
+            app.SetData(RevenantVisuals.Harvesting, false);
+            app.SetData(RevenantVisuals.Stunned, false);
+        }
+
+        //give ghost vis flags
+        var visibility = EntityManager.EnsureComponent<VisibilityComponent>(component.Owner);
+
+        _visibility.AddLayer(visibility, (int) VisibilityFlags.Ghost, false);
+        _visibility.RemoveLayer(visibility, (int) VisibilityFlags.Normal, false);
+        _visibility.RefreshVisibility(visibility);
+
+        //ghost vision
+        if (TryComp(component.Owner, out EyeComponent? eye))
+            eye.VisibilityMask |= (uint) (VisibilityFlags.Ghost);
+
+        //get all the abilities
+        foreach (var listing in _proto.EnumeratePrototypes<RevenantStoreListingPrototype>())
+            component.Listings.Add(listing);
```

Comment:
If you're caching prototypes you should also support prototype hotloading, just need to subscribe to the action on prototypemanager and refresh them.